builtInFunctions/setGuardian.go

```diff
@@ -0,0 +1,270 @@
+		return nil, err
+	}
+	if sg.contains(guardians, vmInput.Arguments[0]) {
+		return nil, ErrGuardianAlreadyExists
+	}
+
+	switch len(guardians.Data) {
+	case 0:
+		// Case 1
+		return sg.tryAddGuardian(senderAccount, vmInput.Arguments[0], guardians, vmInput.GasProvided)
+	case 1:
+		// Case 2
+		if sg.pending(guardians.Data[0]) {
+			return nil, fmt.Errorf("%w: %s", ErrOwnerAlreadyHasOneGuardianPending, sg.pubKeyConverter.Encode(guardians.Data[0].Address))
+		}
+		// Case 3
+		return sg.tryAddGuardian(senderAccount, vmInput.Arguments[0], guardians, vmInput.GasProvided)
+	case 2:
+		// Case 4
+		if sg.pending(guardians.Data[1]) {
+			return nil, fmt.Errorf("%w: %s", ErrOwnerAlreadyHasOneGuardianPending, sg.pubKeyConverter.Encode(guardians.Data[1].Address))
+		}
+		// Case 5
+		guardians.Data = guardians.Data[1:] // remove oldest guardian
+		return sg.tryAddGuardian(senderAccount, vmInput.Arguments[0], guardians, vmInput.GasProvided)
+	default:
+		return &vmcommon.VMOutput{ReturnCode: vmcommon.ExecutionFailed}, nil
+	}
+}
+
+func (sg *setGuardian) checkArguments(
+	senderAccount,
+	receiverAccount vmcommon.UserAccountHandler,
+	vmInput *vmcommon.ContractCallInput,
+) error {
+	if check.IfNil(senderAccount) {
+		return fmt.Errorf("%w for sender", ErrNilUserAccount)
+	}
+	if check.IfNil(receiverAccount) {
+		return fmt.Errorf("%w for receiver", ErrNilUserAccount)
+	}
+	if vmInput == nil {
+		return ErrNilVmInput
+	}
+	if !(bytes.Equal(senderAccount.AddressBytes(), receiverAccount.AddressBytes()) &&
+		bytes.Equal(senderAccount.AddressBytes(), vmInput.CallerAddr)) {
+		return ErrOperationNotPermitted
+	}
+	if vmInput.CallValue == nil {
+		return ErrNilValue
+	}
+	if !isZero(vmInput.CallValue) {
+		return ErrBuiltInFunctionCalledWithValue
+	}
+	if len(vmInput.Arguments) != 1 {
+		return fmt.Errorf("%w, expected 1, got %d ", ErrInvalidNumberOfArguments, len(vmInput.Arguments))
+	}
+	if !sg.isAddressValid(vmInput.Arguments[0]) {
+		return fmt.Errorf("%w for guardian", ErrInvalidAddress)
+	}
+	if bytes.Equal(vmInput.CallerAddr, vmInput.Arguments[0]) {
+		return ErrCannotOwnAddressAsGuardian
+	}
+	if vmInput.GasProvided < sg.funcGasCost {
+		return ErrNotEnoughGas
+	}
+
+	return nil
+}
+
+func isZero(n *big.Int) bool {
+	return len(n.Bits()) == 0
+}
+
+func (sg *setGuardian) isAddressValid(addressBytes []byte) bool {
+	isLengthOk := len(addressBytes) == sg.pubKeyConverter.Len()
+	if !isLengthOk {
+		return false
+	}
+
+	encodedAddress := sg.pubKeyConverter.Encode(addressBytes)
+	return encodedAddress != ""
+}
+
+func (sg *setGuardian) guardians(account vmcommon.UserAccountHandler) (*Guardians, error) {
+	marshalledData, err := account.AccountDataHandler().RetrieveValue(sg.keyPrefix)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fine, account has no guardian set
+	if len(marshalledData) == 0 {
+		return &Guardians{Data: make([]*Guardian, 0)}, nil
+	}
+
+	guardians := &Guardians{}
+	err = sg.marshaller.Unmarshal(guardians, marshalledData)
```

Comment:
if err != nil {
return nil, err
}

Reply:
If `Unmarshal` failed, it would return an error anyway, but indeed it would not return `&Guardians{} == nil`
Refactored it as your suggestion + added extra unit test